core/src/apps/bitcoin/verify_message.py

```diff
@@ -18,9 +20,44 @@
 )
 
 if TYPE_CHECKING:
+    from apps.common.coininfo import CoinInfo
     from trezor.messages import VerifyMessage
 
 
+def address_to_script_type(address: str, coin: CoinInfo) -> InputScriptType:
+    # Determines the script type from a non-multisig address.
+
+    if coin.bech32_prefix and address.startswith(coin.bech32_prefix):
+        witver, _ = common.decode_bech32_address(coin.bech32_prefix, address)
+        if witver == 0:
+            return InputScriptType.SPENDWITNESS
+        elif witver == 1:
+            return InputScriptType.SPENDTAPROOT
+        else:
+            raise wire.DataError("Invalid address")
+
+    if (
+        not utils.BITCOIN_ONLY
+        and coin.cashaddr_prefix is not None
+        and address.startswith(coin.cashaddr_prefix + ":")
+    ):
+        return InputScriptType.SPENDADDRESS
+
+    try:
+        raw_address = base58.decode_check(address, coin.b58_hash)
+    except ValueError:
+        raise wire.DataError("Invalid address")
+
+    if address_type.check(coin.address_type, raw_address):
+        # p2pkh
+        return InputScriptType.SPENDADDRESS
+    elif address_type.check(coin.address_type_p2sh, raw_address):
+        # p2sh
+        return InputScriptType.SPENDP2SHWITNESS
+
+    raise wire.DataError("Invalid address")
```

Comment:
small nit: in case the user passes an invalid address and an undetermined signature, they get `DataError` "Invalid address".
if they pass in an invalid address but one of the specific signatures, they get `ProcessError` "Invalid signature" (from recovered pkey != address)

To do this fully correctly, we would need to parse the address in all cases and then compare the `recid` with what we determined.
But I'm OK with leaving this as-is, just wanted to raise the point in case you see some sort of problem with it.

Reply:
Yes. I meant, we parse address first (to report "malformed address" condition), then fold all subsequent errors (such as recid mismatch) into "invalid signature" condition.

It would not be too complicate to do the same on T1 -- `EXTERNAL` means "bad address".

I am still not sure if it's worth doing. But ISTM it is a good idea to change T1's `DataError` to `ProcessError` to unify the behavior -- invalid signature is not an error in data.